File: python/engines/queue/queue.py
import os
import json
import uuid
import logging
import asyncio
from typing import Dict, Any, Optional, List, Callable
from datetime import datetime, timezone, timedelta
from dataclasses import dataclass, asdict

logger = logging.getLogger("equilibrium.queue")

# Optional Redis import
try:
    import redis.asyncio as redis
    REDIS_AVAILABLE = True
except ImportError:
    REDIS_AVAILABLE = False

@dataclass
class QueuedWorkflow:
    """Workflow queued for processing"""
    id: str
    user_id: str
    task: str
    agents: List[str]
    budget: Dict[str, Any]
    mode: str = "agents"
    meta: Dict[str, Any] = None
    priority: int = 0
    retry_count: int = 0
    max_retries: int = 3
    created_at: float = None
    
    def __post_init__(self):
        if self.created_at is None:
            self.created_at = datetime.now(timezone.utc).timestamp()
        if self.meta is None:
            self.meta = {}

@dataclass
class WorkerHeartbeat:
    """Worker health monitoring"""
    worker_id: str
    status: str  # idle, busy, error
    current_workflow: Optional[str] = None
    last_seen: float = None
    capabilities: List[str] = None
    
    def __post_init__(self):
        if self.last_seen is None:
            self.last_seen = datetime.now(timezone.utc).timestamp()
        if self.capabilities is None:
            self.capabilities = ["agents", "builder"]

class RedisQueueManager:
    """Redis-based distributed queue with seamless in-memory fallback for local environments."""
    
    def __init__(self, redis_url: Optional[str] = None):
        self.redis_url = redis_url or os.getenv("REDIS_URL") or "redis://localhost:6379/0"
        self.redis = None
        self.worker_id = str(uuid.uuid4())
        self.use_fallback = not REDIS_AVAILABLE
        
        # In-memory fallbacks
        self._fallback_queue: asyncio.Queue = asyncio.Queue()
        self._fallback_priority_queue: asyncio.Queue = asyncio.Queue()
        self._fallback_processing: Dict[str, Dict[str, Any]] = {}
        self._fallback_completed: Dict[str, Dict[str, Any]] = {}
        self._fallback_failed: Dict[str, Dict[str, Any]] = {}
        self._fallback_workers: Dict[str, Dict[str, Any]] = {}
        self._fallback_heartbeats: Dict[str, Dict[str, Any]] = {}
        self._fallback_sse_connections: Dict[str, Dict[str, Any]] = {}
        self._fallback_user_sets: Dict[str, set] = {}
        self._fallback_subscribers: List[Callable] = []

        # Key registries
        self.workflow_queue = "eq:workflows:pending"
        self.priority_queue = "eq:workflows:priority"
        self.processing_queue = "eq:workflows:processing"
        self.completed_queue = "eq:workflows:completed"
        self.failed_queue = "eq:workflows:failed"
        self.sse_channels = "eq:sse:channels"
        self.workflow_updates = "eq:workflow:updates"
        self.workers_registry = "eq:workers:registry"
        self.worker_heartbeats = "eq:workers:heartbeats"

    async def connect(self):
        """Connect to Redis. Fallback to in-memory on failure or missing packages."""
        if self.use_fallback:
            return
            
        if not self.redis:
            try:
                self.redis = redis.from_url(
                    self.redis_url,
                    decode_responses=True,
                    socket_connect_timeout=5
                )
                await self.redis.ping()
                logger.info(f"Connected to Redis queue: {self.redis_url}")
            except Exception as e:
                logger.warning(f"Redis connection failed ({e}). Falling back to fully functional in-memory queue.")
                self.use_fallback = True
                self.redis = None

    async def disconnect(self):
        if self.redis:
            await self.redis.close()
            self.redis = None

    async def enqueue_workflow(self, workflow: QueuedWorkflow, high_priority: bool = False) -> str:
        await self.connect()
        
        if self.use_fallback:
            workflow_data = asdict(workflow)
            if high_priority:
                await self._fallback_priority_queue.put(workflow_data)
            else:
                await self._fallback_queue.put(workflow_data)
                
            # Trigger updates to subscriber callbacks
            event = {
                "type": "workflow_queued",
                "workflow_id": workflow.id,
                "user_id": workflow.user_id,
                "queue": self.priority_queue if high_priority else self.workflow_queue
            }
            for cb in self._fallback_subscribers:
                asyncio.create_task(cb(event))
            return workflow.id

        # Redis path
        workflow_data = json.dumps(asdict(workflow))
        queue_name = self.priority_queue if high_priority else self.workflow_queue
        score = workflow.priority if high_priority else 0
        await self.redis.zadd(queue_name, {workflow_data: score})
        await self.redis.publish(
            self.workflow_updates,
            json.dumps({
                "type": "workflow_queued",
                "workflow_id": workflow.id,
                "user_id": workflow.user_id,
                "queue": queue_name
            })
        )
        return workflow.id

    async def dequeue_workflow(self, timeout: int = 5) -> Optional[QueuedWorkflow]:
        await self.connect()
        
        if self.use_fallback:
            try:
                # Try priority queue first
                if not self._fallback_priority_queue.empty():
                    item = await self._fallback_priority_queue.get()
                else:
                    # Async wait for timeout
                    item = await asyncio.wait_for(self._fallback_queue.get(), timeout=float(timeout))
                
                workflow = QueuedWorkflow(**item)
                await self.mark_workflow_processing(workflow)
                return workflow
            except asyncio.TimeoutError:
                return None
                
        # Redis path
        queues = [self.priority_queue, self.workflow_queue]
        for queue in queues:
            result = await self.redis.bzpopmax(queue, timeout=timeout)
            if result:
                _, workflow_data, _ = result
                workflow = QueuedWorkflow(**json.loads(workflow_data))
                await self.mark_workflow_processing(workflow)
                return workflow
        return None

    async def mark_workflow_processing(self, workflow: QueuedWorkflow):
        await self.connect()
        processing_data = {
            "workflow": asdict(workflow),
            "worker_id": self.worker_id,
            "started_at": datetime.now(timezone.utc).timestamp()
        }
        
        if self.use_fallback:
            self._fallback_processing[workflow.id] = processing_data
            event = {
                "type": "workflow_started",
                "workflow_id": workflow.id,
                "worker_id": self.worker_id
            }
            for cb in self._fallback_subscribers:
                asyncio.create_task(cb(event))
            return

        await self.redis.hset(self.processing_queue, workflow.id, json.dumps(processing_data))
        await self.redis.publish(
            self.workflow_updates,
            json.dumps({
                "type": "workflow_started",
                "workflow_id": workflow.id,
                "worker_id": self.worker_id
            })
        )

    async def mark_workflow_completed(self, workflow_id: str, result: Dict[str, Any]):
        await self.connect()
        completed_data = {
            "workflow_id": workflow_id,
            "worker_id": self.worker_id,
            "result": result,
            "completed_at": datetime.now(timezone.utc).timestamp()
        }
        
        if self.use_fallback:
            self._fallback_processing.pop(workflow_id, None)
            self._fallback_completed[workflow_id] = completed_data
            event = {
                "type": "workflow_completed",
                "workflow_id": workflow_id,
                "worker_id": self.worker_id,
                "result": result
            }
            for cb in self._fallback_subscribers:
                asyncio.create_task(cb(event))
            return

        await self.redis.hdel(self.processing_queue, workflow_id)
        await self.redis.hset(self.completed_queue, workflow_id, json.dumps(completed_data))
        await self.redis.publish(
            self.workflow_updates,
            json.dumps({
                "type": "workflow_completed",
                "workflow_id": workflow_id,
                "worker_id": self.worker_id,
                "result": result
            })
        )

    async def mark_workflow_failed(self, workflow_id: str, error: str, retry: bool = True):
        await self.connect()
        
        workflow = None
        if self.use_fallback:
            proc_info = self._fallback_processing.get(workflow_id)
            if proc_info:
                workflow = QueuedWorkflow(**proc_info["workflow"])
        else:
            proc_data = await self.redis.hget(self.processing_queue, workflow_id)
            if proc_data:
                workflow = QueuedWorkflow(**json.loads(proc_data)["workflow"])

        if not workflow:
            return

        will_retry = retry and workflow.retry_count < workflow.max_retries
        if will_retry:
            workflow.retry_count += 1
            delay = min(300, 10 * (2 ** workflow.retry_count))
            logger.info(f"Re-queueing workflow {workflow.id} in {delay}s (Attempt {workflow.retry_count})")
            
            # Run delayed retry task in background
            async def delayed_requeue():
                await asyncio.sleep(delay)
                await self.enqueue_workflow(workflow)
            asyncio.create_task(delayed_requeue())
        else:
            failed_data = {
                "workflow": asdict(workflow),
                "worker_id": self.worker_id,
                "error": error,
                "failed_at": datetime.now(timezone.utc).timestamp()
            }
            if self.use_fallback:
                self._fallback_failed[workflow_id] = failed_data
            else:
                await self.redis.hset(self.failed_queue, workflow_id, json.dumps(failed_data))

        # Remove from processing
        if self.use_fallback:
            self._fallback_processing.pop(workflow_id, None)
            event = {
                "type": "workflow_failed",
                "workflow_id": workflow_id,
                "worker_id": self.worker_id,
                "error": error,
                "retry_count": workflow.retry_count,
                "will_retry": will_retry
            }
            for cb in self._fallback_subscribers:
                asyncio.create_task(cb(event))
            return

        await self.redis.hdel(self.processing_queue, workflow_id)
        await self.redis.publish(
            self.workflow_updates,
            json.dumps({
                "type": "workflow_failed",
                "workflow_id": workflow_id,
                "worker_id": self.worker_id,
                "error": error,
                "retry_count": workflow.retry_count,
                "will_retry": will_retry
            })
        )

    async def register_worker(self, capabilities: List[str] = None):
        await self.connect()
        heartbeat = WorkerHeartbeat(
            worker_id=self.worker_id,
            status="idle",
            capabilities=capabilities or ["agents", "builder"]
        )
        
        if self.use_fallback:
            self._fallback_workers[self.worker_id] = asdict(heartbeat)
            return

        await self.redis.hset(self.workers_registry, self.worker_id, json.dumps(asdict(heartbeat)))

    async def update_worker_heartbeat(self, status: str = "idle", current_workflow: str = None):
        await self.connect()
        heartbeat_data = {
            "worker_id": self.worker_id,
            "status": status,
            "current_workflow": current_workflow,
            "last_seen": datetime.now(timezone.utc).timestamp()
        }
        
        if self.use_fallback:
            self._fallback_heartbeats[self.worker_id] = heartbeat_data
            if self.worker_id in self._fallback_workers:
                self._fallback_workers[self.worker_id].update(heartbeat_data)
            return

        await self.redis.hset(self.worker_heartbeats, self.worker_id, json.dumps(heartbeat_data))

    async def get_queue_stats(self) -> Dict[str, Any]:
        await self.connect()
        if self.use_fallback:
            return {
                "pending_workflows": self._fallback_queue.qsize(),
                "priority_workflows": self._fallback_priority_queue.qsize(),
                "processing_workflows": len(self._fallback_processing),
                "completed_workflows": len(self._fallback_completed),
                "failed_workflows": len(self._fallback_failed),
                "active_workers": len(self._fallback_workers),
                "total_sse_connections": len(self._fallback_sse_connections)
            }
            
        return {
            "pending_workflows": await self.redis.zcard(self.workflow_queue),
            "priority_workflows": await self.redis.zcard(self.priority_queue),
            "processing_workflows": await self.redis.hlen(self.processing_queue),
            "completed_workflows": await self.redis.hlen(self.completed_queue),
            "failed_workflows": await self.redis.hlen(self.failed_queue),
            "active_workers": await self.redis.hlen(self.workers_registry),
            "total_sse_connections": await self.redis.hlen(f"{self.sse_channels}:connections")
        }

    async def register_sse_connection(self, connection_id: str, user_id: str):
        await self.connect()
        connection_data = {
            "connection_id": connection_id,
            "user_id": user_id,
            "worker_id": self.worker_id,
            "connected_at": datetime.now(timezone.utc).timestamp()
        }
        
        if self.use_fallback:
            self._fallback_sse_connections[connection_id] = connection_data
            if user_id not in self._fallback_user_sets:
                self._fallback_user_sets[user_id] = set()
            self._fallback_user_sets[user_id].add(connection_id)
            return

        await self.redis.hset(f"{self.sse_channels}:connections", connection_id, json.dumps(connection_data))
        await self.redis.sadd(f"{self.sse_channels}:user:{user_id}", connection_id)

    async def unregister_sse_connection(self, connection_id: str, user_id: str):
        await self.connect()
        if self.use_fallback:
            self._fallback_sse_connections.pop(connection_id, None)
            if user_id in self._fallback_user_sets:
                self._fallback_user_sets[user_id].discard(connection_id)
            return

        await self.redis.hdel(f"{self.sse_channels}:connections", connection_id)
        await self.redis.srem(f"{self.sse_channels}:user:{user_id}", connection_id)

    async def subscribe_to_workflow_updates(self, callback: Callable):
        await self.connect()
        if self.use_fallback:
            self._fallback_subscribers.append(callback)
            return

        pubsub = self.redis.pubsub()
        await pubsub.subscribe(self.workflow_updates)
        
        async def listen_loop():
            try:
                async for message in pubsub.listen():
                    if message["type"] == "message":
                        try:
                            event_data = json.loads(message["data"])
                            await callback(event_data)
                        except Exception as e:
                            logger.error(f"Error processing workflow update: {e}")
            except Exception as e:
                logger.error(f"Workflow update subscription channel error: {e}")
                
        asyncio.create_task(listen_loop())

File: esm/engines/04_queue/queue.js
/**
 * Redis-based distributed queue for workflow processing.
 * Upgraded to ESModules with a robust in-memory event-driven fallback.
 */

import crypto from 'crypto';
import EventEmitter from 'events';

let redisModule = null;
let useFallback = false;

// Attempt to load ioredis or redis dynamically
try {
  const ioRedis = await import('ioredis');
  redisModule = ioRedis.default;
} catch (_) {
  try {
    const redisPkg = await import('redis');
    redisModule = redisPkg;
  } catch (err) {
    console.warn(`[QueueManager] No Redis packages found (ioredis/redis). Defaulting to in-memory queue fallback.`);
    useFallback = true;
  }
}

export class QueuedWorkflow {
  constructor({ id, userId, task, agents, budget, mode = 'agents', meta = {}, priority = 0, retryCount = 0, maxRetries = 3 }) {
    this.id = id || crypto.randomUUID();
    this.userId = userId;
    this.task = task;
    this.agents = agents;
    this.budget = budget;
    this.mode = mode;
    this.meta = meta;
    this.priority = priority;
    this.retryCount = retryCount;
    this.maxRetries = maxRetries;
    this.createdAt = Date.now() / 1000;
  }

  toObject() {
    return {
      id: this.id,
      userId: this.userId,
      task: this.task,
      agents: this.agents,
      budget: this.budget,
      mode: this.mode,
      meta: this.meta,
      priority: this.priority,
      retryCount: this.retryCount,
      maxRetries: this.maxRetries,
      createdAt: this.createdAt
    };
  }
}

export class RedisQueueManager {
  constructor(redisUrl = null) {
    this.redisUrl = redisUrl || process.env.REDIS_URL || 'redis://localhost:6379/0';
    this.client = null;
    this.workerId = crypto.randomUUID();
    this.useFallback = useFallback;

    // In-memory queues and tables
    this._fallbackQueue = [];
    this._fallbackPriorityQueue = [];
    this._fallbackProcessing = new Map();
    this._fallbackCompleted = new Map();
    this._fallbackFailed = new Map();
    this._fallbackWorkers = new Map();
    this._fallbackHeartbeats = new Map();
    this._fallbackSseConnections = new Map();
    this._fallbackUserSets = new Map();
    this._fallbackEmitter = new EventEmitter();

    // Registry keys
    this.workflowQueue = "eq:workflows:pending";
    this.priorityQueue = "eq:workflows:priority";
    this.processingQueue = "eq:workflows:processing";
    this.completedQueue = "eq:workflows:completed";
    this.failedQueue = "eq:workflows:failed";
    this.sseChannels = "eq:sse:channels";
    this.workflowUpdates = "eq:workflow:updates";
    this.workersRegistry = "eq:workers:registry";
    this.workerHeartbeats = "eq:workers:heartbeats";
  }

  async connect() {
    if (this.useFallback) return;

    if (!this.client) {
      try {
        if (redisModule) {
          // If using ioredis
          if (typeof redisModule === 'function') {
            this.client = new redisModule(this.redisUrl);
          } else if (redisModule.createClient) {
            this.client = redisModule.createClient({ url: this.redisUrl });
            await this.client.connect();
          }
          // Validate connection
          await this.client.ping();
          console.log(`[QueueManager] Connected to Redis queue: ${this.redisUrl}`);
        } else {
          this.useFallback = true;
        }
      } catch (err) {
        console.warn(`[QueueManager] Redis connection failed (${err.message}). Falling back to in-memory queue.`);
        this.useFallback = true;
        this.client = null;
      }
    }
  }

  async disconnect() {
    if (this.client) {
      if (typeof this.client.quit === 'function') {
        await this.client.quit();
      } else if (typeof this.client.disconnect === 'function') {
        await this.client.disconnect();
      }
      this.client = null;
    }
  }

  async enqueueWorkflow(workflow, highPriority = false) {
    await this.connect();
    const data = workflow.toObject();

    if (this.useFallback) {
      if (highPriority) {
        // Sort insertion by priority score
        this._fallbackPriorityQueue.push(data);
        this._fallbackPriorityQueue.sort((a, b) => b.priority - a.priority);
      } else {
        this._fallbackQueue.push(data);
      }

      const event = {
        type: "workflow_queued",
        workflowId: workflow.id,
        userId: workflow.userId,
        queue: highPriority ? this.priorityQueue : this.workflowQueue
      };
      this._fallbackEmitter.emit(this.workflowUpdates, event);
      return workflow.id;
    }

    // Redis flow (using sorted set)
    const payload = JSON.stringify(data);
    const queueName = highPriority ? this.priorityQueue : this.workflowQueue;
    const score = highPriority ? workflow.priority : 0;
    
    if (typeof this.client.zadd === 'function') {
      await this.client.zadd(queueName, score, payload);
      await this.client.publish(this.workflowUpdates, JSON.stringify({
        type: "workflow_queued",
        workflowId: workflow.id,
        userId: workflow.userId,
        queue: queueName
      }));
    } else {
      // redis v4 format
      await this.client.zAdd(queueName, { score, value: payload });
      await this.client.publish(this.workflowUpdates, JSON.stringify({
        type: "workflow_queued",
        workflowId: workflow.id,
        userId: workflow.userId,
        queue: queueName
      }));
    }
    return workflow.id;
  }

  async dequeueWorkflow(timeout = 5) {
    await this.connect();

    if (this.useFallback) {
      if (this._fallbackPriorityQueue.length > 0) {
        const item = this._fallbackPriorityQueue.shift();
        const workflow = new QueuedWorkflow(item);
        await this.markWorkflowProcessing(workflow);
        return workflow;
      }
      if (this._fallbackQueue.length > 0) {
        const item = this._fallbackQueue.shift();
        const workflow = new QueuedWorkflow(item);
        await this.markWorkflowProcessing(workflow);
        return workflow;
      }
      // Wait for item using standard event emitter with timeout
      return new Promise((resolve) => {
        const timer = setTimeout(() => {
          this._fallbackEmitter.off('new_item', handler);
          resolve(null);
        }, timeout * 1000);

        const handler = (item) => {
          clearTimeout(timer);
          const workflow = new QueuedWorkflow(item);
          this.markWorkflowProcessing(workflow).then(() => resolve(workflow));
        };

        this._fallbackEmitter.once('new_item', handler);
      });
    }

    // Redis flow
    const queues = [this.priorityQueue, this.workflowQueue];
    for (const queue of queues) {
      let result = null;
      if (typeof this.client.bzpopmax === 'function') {
        result = await this.client.bzpopmax(queue, timeout);
      } else {
        // node-redis v4
        result = await this.client.bzPopMax(queue, timeout);
      }
      
      if (result) {
        // bzpopmax returns [key, member, score] or {key, element, score}
        const workflowData = Array.isArray(result) ? result[1] : (result.element || result.value);
        const workflow = new QueuedWorkflow(JSON.parse(workflowData));
        await this.markWorkflowProcessing(workflow);
        return workflow;
      }
    }
    return null;
  }

  async markWorkflowProcessing(workflow) {
    await this.connect();
    const data = {
      workflow: workflow.toObject(),
      workerId: this.workerId,
      startedAt: Date.now() / 1000
    };

    if (this.useFallback) {
      this._fallbackProcessing.set(workflow.id, data);
      this._fallbackEmitter.emit(this.workflowUpdates, {
        type: "workflow_started",
        workflowId: workflow.id,
        workerId: this.workerId
      });
      return;
    }

    const payload = JSON.stringify(data);
    if (typeof this.client.hset === 'function') {
      await this.client.hset(this.processingQueue, workflow.id, payload);
      await this.client.publish(this.workflowUpdates, JSON.stringify({
        type: "workflow_started",
        workflowId: workflow.id,
        workerId: this.workerId
      }));
    } else {
      await this.client.hSet(this.processingQueue, workflow.id, payload);
      await this.client.publish(this.workflowUpdates, JSON.stringify({
        type: "workflow_started",
        workflowId: workflow.id,
        workerId: this.workerId
      }));
    }
  }

  async markWorkflowCompleted(workflowId, result) {
    await this.connect();
    const completedData = {
      workflowId,
      workerId: this.workerId,
      result,
      completedAt: Date.now() / 1000
    };

    if (this.useFallback) {
      this._fallbackProcessing.delete(workflowId);
      this._fallbackCompleted.set(workflowId, completedData);
      this._fallbackEmitter.emit(this.workflowUpdates, {
        type: "workflow_completed",
        workflowId,
        workerId: this.workerId,
        result
      });
      return;
    }

    const payload = JSON.stringify(completedData);
    if (typeof this.client.hdel === 'function') {
      await this.client.hdel(this.processingQueue, workflowId);
      await this.client.hset(this.completedQueue, workflowId, payload);
      await this.client.publish(this.workflowUpdates, JSON.stringify({
        type: "workflow_completed",
        workflowId,
        workerId: this.workerId,
        result
      }));
    } else {
      await this.client.hDel(this.processingQueue, workflowId);
      await this.client.hSet(this.completedQueue, workflowId, payload);
      await this.client.publish(this.workflowUpdates, JSON.stringify({
        type: "workflow_completed",
        workflowId,
        workerId: this.workerId,
        result
      }));
    }
  }

  async markWorkflowFailed(workflowId, error, retry = true) {
    await this.connect();
    let workflow = null;

    if (this.useFallback) {
      const procInfo = this._fallbackProcessing.get(workflowId);
      if (procInfo) workflow = new QueuedWorkflow(procInfo.workflow);
    } else {
      const procData = typeof this.client.hget === 'function' 
        ? await this.client.hget(this.processingQueue, workflowId)
        : await this.client.hGet(this.processingQueue, workflowId);
      if (procData) {
        workflow = new QueuedWorkflow(JSON.parse(procData).workflow);
      }
    }

    if (!workflow) return;

    const willRetry = retry && workflow.retryCount < workflow.maxRetries;
    if (willRetry) {
      workflow.retryCount += 1;
      const delay = Math.min(300, 10 * Math.pow(2, workflow.retryCount));
      console.log(`[QueueManager] Re-queueing failing workflow ${workflow.id} in ${delay}s`);
      
      setTimeout(() => {
        this.enqueueWorkflow(workflow);
      }, delay * 1000);
    } else {
      const failedData = {
        workflow: workflow.toObject(),
        workerId: this.workerId,
        error,
        failedAt: Date.now() / 1000
      };
      if (this.useFallback) {
        this._fallbackFailed.set(workflowId, failedData);
      } else {
        const payload = JSON.stringify(failedData);
        if (typeof this.client.hset === 'function') {
          await this.client.hset(this.failedQueue, workflowId, payload);
        } else {
          await this.client.hSet(this.failedQueue, workflowId, payload);
        }
      }
    }

    if (this.useFallback) {
      this._fallbackProcessing.delete(workflowId);
      this._fallbackEmitter.emit(this.workflowUpdates, {
        type: "workflow_failed",
        workflowId,
        workerId: this.workerId,
        error,
        retryCount: workflow.retryCount,
        willRetry
      });
      return;
    }

    if (typeof this.client.hdel === 'function') {
      await this.client.hdel(this.processingQueue, workflowId);
    } else {
      await this.client.hDel(this.processingQueue, workflowId);
    }

    await this.client.publish(this.workflowUpdates, JSON.stringify({
      type: "workflow_failed",
      workflowId,
      workerId: this.workerId,
      error,
      retryCount: workflow.retryCount,
      willRetry
    }));
  }

  async registerWorker(capabilities = null) {
    await this.connect();
    const data = {
      workerId: this.workerId,
      status: "idle",
      capabilities: capabilities || ["agents", "builder"],
      lastSeen: Date.now() / 1000
    };

    if (this.useFallback) {
      this._fallbackWorkers.set(this.workerId, data);
      return;
    }

    const payload = JSON.stringify(data);
    if (typeof this.client.hset === 'function') {
      await this.client.hset(this.workersRegistry, this.workerId, payload);
    } else {
      await this.client.hSet(this.workersRegistry, this.workerId, payload);
    }
  }

  async updateWorkerHeartbeat(status = "idle", currentWorkflow = null) {
    await this.connect();
    const data = {
      workerId: this.workerId,
      status,
      currentWorkflow,
      lastSeen: Date.now() / 1000
    };

    if (this.useFallback) {
      this._fallbackHeartbeats.set(this.workerId, data);
      const wrk = this._fallbackWorkers.get(this.workerId);
      if (wrk) {
        Object.assign(wrk, data);
      }
      return;
    }

    const payload = JSON.stringify(data);
    if (typeof this.client.hset === 'function') {
      await this.client.hset(this.workerHeartbeats, this.workerId, payload);
    } else {
      await this.client.hSet(this.workerHeartbeats, this.workerId, payload);
    }
  }

  async getQueueStats() {
    await this.connect();
    if (this.useFallback) {
      return {
        pendingWorkflows: this._fallbackQueue.length,
        priorityWorkflows: this._fallbackPriorityQueue.length,
        processingWorkflows: this._fallbackProcessing.size,
        completedWorkflows: this._fallbackCompleted.size,
        failedWorkflows: this._fallbackFailed.size,
        activeWorkers: this._fallbackWorkers.size,
        totalSseConnections: this._fallbackSseConnections.size
      };
    }

    if (typeof this.client.zcard === 'function') {
      return {
        pendingWorkflows: await this.client.zcard(this.workflowQueue),
        priorityWorkflows: await this.client.zcard(this.priorityQueue),
        processingWorkflows: await this.client.hlen(self.processingQueue),
        completedWorkflows: await this.client.hlen(this.completedQueue),
        failedWorkflows: await this.client.hlen(this.failedQueue),
        activeWorkers: await this.client.hlen(this.workersRegistry),
        totalSseConnections: await this.client.hlen(`${this.sseChannels}:connections`)
      };
    } else {
      return {
        pendingWorkflows: await this.client.zCard(this.workflowQueue),
        priorityWorkflows: await this.client.zCard(this.priorityQueue),
        processingWorkflows: await this.client.hLen(this.processingQueue),
        completedWorkflows: await this.client.hLen(this.completedQueue),
        failedWorkflows: await this.client.hLen(this.failedQueue),
        activeWorkers: await this.client.hLen(this.workersRegistry),
        totalSseConnections: await this.client.hLen(`${this.sseChannels}:connections`)
      };
    }
  }

  async registerSseConnection(connectionId, userId) {
    await this.connect();
    const data = {
      connectionId,
      userId,
      workerId: this.workerId,
      connectedAt: Date.now() / 1000
    };

    if (this.useFallback) {
      this._fallbackSseConnections.set(connectionId, data);
      if (!this._fallbackUserSets.has(userId)) {
        this._fallbackUserSets.set(userId, new Set());
      }
      this._fallbackUserSets.get(userId).add(connectionId);
      return;
    }

    const payload = JSON.stringify(data);
    if (typeof this.client.hset === 'function') {
      await this.client.hset(`${this.sseChannels}:connections`, connectionId, payload);
      await this.client.sadd(`${this.sseChannels}:user:${userId}`, connectionId);
    } else {
      await this.client.hSet(`${this.sseChannels}:connections`, connectionId, payload);
      await this.client.sAdd(`${this.sseChannels}:user:${userId}`, connectionId);
    }
  }

  async unregisterSseConnection(connectionId, userId) {
    await this.connect();
    if (this.useFallback) {
      this._fallbackSseConnections.delete(connectionId);
      if (this._fallbackUserSets.has(userId)) {
        this._fallbackUserSets.get(userId).delete(connectionId);
      }
      return;
    }

    if (typeof this.client.hdel === 'function') {
      await this.client.hdel(`${this.sseChannels}:connections`, connectionId);
      await this.client.srem(`${this.sseChannels}:user:${userId}`, connectionId);
    } else {
      await this.client.hDel(`${this.sseChannels}:connections`, connectionId);
      await this.client.sRem(`${this.sseChannels}:user:${userId}`, connectionId);
    }
  }

  async subscribeToWorkflowUpdates(callback) {
    await this.connect();
    if (this.useFallback) {
      this._fallbackEmitter.on(this.workflowUpdates, callback);
      return;
    }

    // In native Node redis, subscription must happen on a separate connection, but for simplicity
    // we use a pubsub subscription mechanism.
    try {
      let subClient = null;
      if (typeof redisModule === 'function') {
        subClient = new redisModule(this.redisUrl);
      } else if (redisModule.createClient) {
        subClient = redisModule.createClient({ url: this.redisUrl });
        await subClient.connect();
      }

      if (typeof subClient.subscribe === 'function') {
        await subClient.subscribe(this.workflowUpdates, (msg) => {
          try {
            callback(JSON.parse(msg));
          } catch (_) {
            callback(msg);
          }
        });
      }
    } catch (e) {
      console.error(`[QueueManager] Subscription failed: ${e.message}`);
    }
  }
}
